### DISCUSSION TITLE
Auto generate query key

### DISCUSSION THREAD

**Discussion by melloc01** [TanStack/query#1873] [Q&A]
I am considering adding `react-query` to a large, several teams project, but I am foreseeing that the query key being manually put will cause problems of key clashing.

Do react-query has plans to provide a way to auto-generate those keys? Something like rtk-query does with URL & params would work well I think.

Having the queries declared `on-use` is a problem I get that, I am just bringing the subject for discussion.

**Comment by TkDodo** [comment]
I think there is a Babel plugin that does exactly that. See: 
- https://mobile.twitter.com/dominictwlee/status/1364972001782341636
- https://github.com/dominictwlee/babel-plugin-react-query-key-gen

**Reply by melloc01** [reply]
That's nice, thanks for the quick response @TkDodo - this is actually helpful since we centralize our babel configuration.

I will give it a try.

Do you know if there are any Feature Requests or "official" plans to get it included in the library?
I couldn't find it, I'd be happy to create it if there's isn't

When it comes to big projects, I guess this will be a recurring issue

**Reply by dominictwlee** [reply]
Thanks for the mention @TkDodo!

@melloc01 Yeah give it a go and see whether it works out well for you. Also feel free to contribute to it if it's missing use cases or functionality that you need! It's relatively new and not battle tested so use with discretion :). Happy to invite you as a collaborator if you're interested.

**Reply by TkDodo** [reply]
> When it comes to big projects, I guess this will be a recurring issue

As usual, "it depends" is the only valid answer :) We are having rather big projects running with react-query, and key collusions are not a factor. Every feature lives in it's own directory and has a key starting with the name of the feature. Yes, it's just a convention, and there are queries used by multiple features, so they are global, but the risk is still very minimal imo.
